crates/turborepo-lib/src/shim.rs

```diff
@@ -306,48 +301,49 @@ impl LocalTurboState {
     // - berry (nodeLinker: "node-modules")
     //
     // This also supports people directly depending upon the platform version.
-    fn generate_hoisted_path(root_path: &Utf8Path) -> Option<Utf8PathBuf> {
-        Some(root_path.join("node_modules"))
+    fn generate_hoisted_path(root_path: &AbsoluteSystemPath) -> Option<AbsoluteSystemPathBuf> {
+        Some(root_path.join_component("node_modules"))
     }
 
     // Nested strategy:
     // - `npm install --install-strategy=shallow` (`npm install --global-style`)
     // - `npm install --install-strategy=nested` (`npm install --legacy-bundling`)
     // - berry (nodeLinker: "pnpm")
-    fn generate_nested_path(root_path: &Utf8Path) -> Option<Utf8PathBuf> {
-        Some(
-            root_path
-                .join("node_modules")
-                .join("turbo")
-                .join("node_modules"),
-        )
+    fn generate_nested_path(root_path: &AbsoluteSystemPath) -> Option<AbsoluteSystemPathBuf> {
+        Some(root_path.join_components(&["node_modules", "turbo", "node_modules"]))
     }
 
     // Linked strategy:
     // - `pnpm install`
     // - `npm install --install-strategy=linked`
-    fn generate_linked_path(root_path: &Utf8Path) -> Option<Utf8PathBuf> {
-        let canonical_path =
-            fs_canonicalize(root_path.join("node_modules").join("turbo").join("..")).ok()?;
+    fn generate_linked_path(root_path: &AbsoluteSystemPath) -> Option<AbsoluteSystemPathBuf> {
+        let canonical_path = fs_canonicalize(
+            root_path
+                .as_path()
+                .join("node_modules")
+                .join("turbo")
+                .join(".."),
```

Comment:
```suggestion
root_path.join_component("node_modules").as_path()
```

Reply:
I originally did the equivalent, but had to walk that back after our tests caught a limitation with that approach:

https://github.com/vercel/turbo/pull/5986/files/721c8bc947b4bef8904a10c009bd0c2828774753..e1252d6bd9ed7952c8968fa872a5770b1eda423e#diff-8459a232d027b507053b9df40feba72a5fdc6b1b08d5348140ab2ce9dd4d861e

Reply:
(`AbsoluteSystemPath` does a lexical clean, underlying system behavior just does a string concat. If we lexically clean prior to file-system-aware canonicalization we get the wrong output.)